### PR TITLE
Experiment: switching @asgrim's much better roave/better-reflection stuff

### PR DESCRIPTION
**Pull request by Ocramius**
Note: still incomplete, as roave/better-reflection didn't reach parity with internals reflection yet.

### CODE REVIEW

File: src/ProxyManager/GeneratorStrategy/FileWriterGeneratorStrategy.php [hunk 1]
```diff
@@ -62,6 +62,8 @@ public function generate(ClassGenerator $classGenerator)
         file_put_contents($tmpFileName, "<?php\n\n" . $generatedCode);
         rename($tmpFileName, $fileName);
 
+        eval($generatedCode);
```
Comment (Ocramius) [inline src/ProxyManager/GeneratorStrategy/FileWriterGeneratorStrategy.php]: To be removed before merge

File: src/ProxyManager/Factory/AbstractBaseFactory.php [hunk 2]
```diff
@@ -112,7 +112,7 @@ private function generateProxyClass($proxyClassName, $className, array $proxyPar
         $className = $this->configuration->getClassNameInflector()->getUserClassName($className);
         $phpClass  = new ClassGenerator($proxyClassName);
 
-        $this->getGenerator()->generate(new ReflectionClass($className), $phpClass, $proxyOptions);
+        $this->getGenerator()->generate(\BetterReflection\Reflection\ReflectionClass::createFromName($className), $phpClass, $proxyOptions);
```
Comment (Ocramius) [inline src/ProxyManager/Factory/AbstractBaseFactory.php]: Requires Roave/BetterReflection#18 to work properly

File: src/ProxyManager/Factory/AbstractBaseFactory.php [hunk 3]
```diff
@@ -87,7 +87,7 @@ protected function generateProxy($className, array $proxyOptions = [])
         $this
             ->configuration
             ->getSignatureChecker()
-            ->checkSignature(new ReflectionClass($proxyClassName), $proxyParameters);
+            ->checkSignature(\BetterReflection\Reflection\ReflectionClass::createFromName($proxyClassName), $proxyParameters);
```
Comment (Ocramius) [inline src/ProxyManager/Factory/AbstractBaseFactory.php]: Requires Roave/BetterReflection#18 to work properly

**Comment by Ocramius** [comment]
Not for 2.0.0: we will likely use this for later versions

**Comment by lisachenko** [comment]
@Ocramius you should also give a try for goaop/parser-reflection library. It's compatible with internal reflection (something like 95% now, only trait adaptation is not implemented yet) And I will work on it for my own dirty needs ;)

**Comment by Ocramius** [comment]
@lisachenko eh, I actively worked with @asgrim on better-reflection, following internal design and such, so while I would love to try your lib, it would probably be much more time consuming to do so instead of just using something that I know very well :-P

**Comment by lisachenko** [comment]
Ok ) I can understand your decision, because we select different goals. I want my library to be as close as possible to the original versions of reflection, so no extra methods and functionality. This should be a basis for top-level libraries...

PS. Looks like some kind of competition will improve our projects ;)

File: src/ProxyManager/GeneratorStrategy/FileWriterGeneratorStrategy.php [hunk 4]
```diff
@@ -74,6 +74,8 @@ public function generate(ClassGenerator $classGenerator)
             restore_error_handler();
         }
 
+        eval($generatedCode);
```
Comment (Ocramius) [inline src/ProxyManager/GeneratorStrategy/FileWriterGeneratorStrategy.php]: to be removed

**Comment by Ocramius** [comment]
> Ok ) I can understand your decision, because we select  different goals. I want my library to be as close as possible to the  original versions of reflection, so no extra methods and functionality.  This should be a basis for top-level libraries...


There is some misunderstanding here: this is a goal also for BetterReflection, except that we have two separate APIs:

 * one for getting "better" reflection
 * one for emulating core reflection (we reached parity there, btw)

So BetterReflection is still a perfect fit.

**Comment by lisachenko** [comment]
> There is some misunderstanding here: this is a goal also for BetterReflection, except that we have two separate APIs:

>    one for getting "better" reflection
>    one for emulating core reflection (we reached parity there, btw)

> So BetterReflection is still a perfect fit.

Of course, it's our choice as maintainers to depend on specific implementation. Because we should have full control over entire project. When I asked in https://github.com/Roave/BetterReflection/issues/18 about inheritance it wasn't ready and (maybe) team of BR hadn't time for this stuff and I couldn't wait any more. So my own implementation was developed from scratch.

I see a lot of interesting stuff in BR, but in my opinion, it violates SRP for the package and tries to do much more than required for simple reflection, eg. AST transformation, searching for class loaders, etc. BR is kind of Swiss knife, but hey, why I need heavy tool? :) I need fast and simple tool that provides me only Reflection API, it should be enough. All other stuff can be implemented on top of this. This is the reason why I still don't use BR.

There are only few things that missing now in goaop/parser-reflection:
  - to be fully transparent for internal classes.
  - to have an option to specify class name for reflection (kind of factory) to have an ability to build anything more complex on top of this library.

At this point I will finish development and only keep it compatible with new versions of PHP. All other stuff should be composed from existing bricks )

So, from what I can see we will keep running both projects.

**Comment by Ocramius** [comment]
> Of course, it's our choice as maintainers to depend on specific implementation. 

I only rely on the reflection API - the only change here (and you can see it from the diffs) is the initial factory/instantiation of a reflector

**Comment by malukenho** [comment]
@Ocramius can this one be closed?

**Comment by Ocramius** [comment]
This one is still valid from a feature perspective, but should be reduced to removing reflection instantiation where possible.

In practice: patch is not valid anymore, but the feature is still wanted, as it allows for proxying stuff like final classes.

**Comment by lisachenko** [comment]
FYI: Proxying of final classes will be available via [lisachenko/z-engine](https://github.com/lisachenko/z-engine) and PHP>=7.4.

```php
use ZEngine\Reflection\ReflectionClass as ReflectionClassEx;

$refClass = new ReflectionClassEx(SomeFinalClass::class);
$refClass->setFinal(false);
```
After that you can extend your final class easily.

**Comment by Ocramius** [comment]
While that is tempting, it would require adding an extension for something that would otherwise require removing the `final` marker by hand (when you **know** what you are doing).

The initial idea of this patch is to generate proxy code without loading the original sources (which may lead to side-effects)

**Pull request closed without merging**